dissect/volume/md/stream.py

```diff
@@ -0,0 +1,394 @@
+from __future__ import annotations
+
+import io
+import operator
+from typing import TYPE_CHECKING, BinaryIO, NamedTuple, Optional
+
+from dissect.util.stream import AlignedStream, MappingStream
+
+from dissect.volume.exceptions import MDError
+from dissect.volume.md.c_md import c_md
+
+if TYPE_CHECKING:
+    from dissect.volume.md.md import MD, Device
+
+
+def create_linear_stream(md: MD) -> BinaryIO:
+    stream = MappingStream(align=md.chunk_size or io.DEFAULT_BUFFER_SIZE)
+
+    offset = 0
+    for device in md.devices:
+        # Ignore special device roles for now
+        if device.raid_disk is None:
+            continue
+
+        device_size = device.data_size * 512
+        stream.add(offset, device_size, device.open(), 0)
+        offset += device_size
+
+    stream.size = offset
+    return stream
+
+
+class Zone(NamedTuple):
+    zone_end: int
+    dev_start: int
+    devices: list[Device]
+
+
+class RAID0Stream(AlignedStream):
+    def __init__(self, md: MD):
+        self.md = md
+        self.devices = {dev.raid_disk: dev for dev in self.md.devices if dev.raid_disk is not None}
+        if len(self.devices) != md.raid_disks:
+            raise MDError(f"Missing disks in RAID0 set {self.md.uuid} ({self.md.name})")
+
+        devices = sorted(self.devices.values(), key=operator.attrgetter("raid_disk"))
+        num_strip_zones = 0
+        for dev1 in devices:
```

Comment:
A bit more comments on the the flow of the different algorithms would help reviewing ;-)

Reply:
There aren't any comments in the original source either so I barely understand how it works myself :smile: